File: iptv.js

const m3uUrl = "https://cdn.jsdelivr.net/gh/PRENDLYMADAPAKER/ANG-KALAT-MO@main/IPTVPREMIUM.m3u";
let channels = [];
let currentIndex = 0;
let video = document.getElementById("videoPlayer");
let overlay = document.getElementById("overlay");
let lastTime = 0;
let stuckCounter = 0;
let reconnecting = false;

// Reconnect overlay element
function showOverlay(show) {
  overlay.style.display = show ? "flex" : "none";
}

// Fetch and parse M3U
async function fetchChannels() {
  try {
    const response = await fetch(m3uUrl);
    const data = await response.text();
    const lines = data.split("\n");
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].startsWith("#EXTINF:")) {
        const nameMatch = lines[i].match(/,(.*)$/);
        const logoMatch = lines[i].match(/tvg-logo="([^"]+)"/);
        const groupMatch = lines[i].match(/group-title="([^"]+)"/);
        const url = lines[i + 1]?.trim();
        if (url && url.startsWith("http")) {
          channels.push({
            name: nameMatch ? nameMatch[1] : "Untitled",
            url,
            logo: logoMatch ? logoMatch[1] : "",
            group: groupMatch ? groupMatch[1] : "Other"
          });
        }
      }
    }
    populateDropdown();
  } catch (err) {
    console.error("Error fetching M3U:", err);
  }
}

// Populate dropdown with parsed channels
function populateDropdown() {
  const select = document.getElementById("channelSelect");
  channels.forEach((ch, index) => {
    const opt = document.createElement("option");
    opt.value = index;
    opt.textContent = ch.name;
    select.appendChild(opt);
  });
}

// Play selected channel
function playChannel(index) {
  if (!channels[index]) return;
  currentIndex = index;
  const ch = channels[index];
  video.src = ch.url;
  document.getElementById("nowPlaying").textContent = ch.name;
  video.load();
  video.play().catch(err => console.error("Playback error:", err));
}

// Monitor video freeze
setInterval(() => {
  if (video.paused || video.readyState < 2 || reconnecting) return;
  if (video.currentTime === lastTime) {
    stuckCounter++;
    if (stuckCounter >= 5) {
      reconnectStream();
    }
  } else {
    stuckCounter = 0;
  }
  lastTime = video.currentTime;
}, 2000);

// Attempt to reconnect the stream
function reconnectStream() {
  if (reconnecting) return;
  reconnecting = true;
  showOverlay(true);
  const ch = channels[currentIndex];
  let attempts = 0;
  const tryReconnect = setInterval(() => {
    if (!reconnecting) {
      clearInterval(tryReconnect);
      return;
    }
    console.log("Retrying stream...");
    video.src = ch.url;
    video.load();
    video.play().then(() => {
      showOverlay(false);
      stuckCounter = 0;
      reconnecting = false;
      clearInterval(tryReconnect);
    }).catch(err => {
      attempts++;
      if (attempts > 10) {
        clearInterval(tryReconnect);
        console.error("Reconnect failed.");
      }
    });
  }, 3000);
}

// UI handlers
document.getElementById("channelSelect").addEventListener("change", e => {
  playChannel(e.target.value);
});

document.getElementById("importBtn").addEventListener("click", () => {
  const customUrl = document.getElementById("customM3u").value;
  if (customUrl) {
    fetch(customUrl).then(res => res.text()).then(data => {
      channels = [];
      document.getElementById("channelSelect").innerHTML = '<option selected disabled>Select a channel</option>';
      const lines = data.split("\n");
      for (let i = 0; i < lines.length; i++) {
        if (lines[i].startsWith("#EXTINF:")) {
          const nameMatch = lines[i].match(/,(.*)$/);
          const logoMatch = lines[i].match(/tvg-logo="([^"]+)"/);
          const groupMatch = lines[i].match(/group-title="([^"]+)"/);
          const url = lines[i + 1]?.trim();
          if (url && url.startsWith("http")) {
            channels.push({
              name: nameMatch ? nameMatch[1] : "Untitled",
              url,
              logo: logoMatch ? logoMatch[1] : "",
              group: groupMatch ? groupMatch[1] : "Other"
            });
          }
        }
      }
      populateDropdown();
    });
  }
});

// Initial load
fetchChannels();

File: iptv.py
const m3uUrl =
  "https://raw.githubusercontent.com/PRENDLYMADAPAKER/ANG-KALAT-MO/refs/heads/main/IPTVPREMIUM.m3u";
let channels = [];
let currentIndex = 0;

const video = document.getElementById("video");
const channelList = document.getElementById("channelList");
const nowPlaying = document.getElementById("nowPlaying");
const searchInput = document.getElementById("searchInput");
const categoryFilter = document.getElementById("categoryFilter");
const importBtn = document.getElementById("importBtn");
const customM3U = document.getElementById("customM3U");

const favorites = JSON.parse(localStorage.getItem("favorites") || "[]");

function setClock() {
  setInterval(() => {
    const now = new Date();
    document.getElementById("clock").textContent = now.toLocaleTimeString();
  }, 1000);
}

function parseM3U(content) {
  const lines = content.split("\n");
  const parsed = [];
  let current = {};

  for (let line of lines) {
    if (line.startsWith("#EXTINF:")) {
      const nameMatch = line.match(/,(.+)$/);
      const logoMatch = line.match(/tvg-logo="(.*?)"/);
      const groupMatch = line.match(/group-title="(.*?)"/);
      current = {
        name: nameMatch ? nameMatch[1].trim() : "Unnamed",
        logo: logoMatch ? logoMatch[1] : "",
        group: groupMatch ? groupMatch[1] : "Others",
        url: "",
      };
    } else if (line.startsWith("http")) {
      current.url = line.trim();
      parsed.push({ ...current });
    }
  }

  return parsed;
}

async function loadM3U(url) {
  try {
    const res = await fetch(url);
    const text = await res.text();
    channels = parseM3U(text);
    renderCategories();
    renderChannels();
    playChannel(0);
  } catch (err) {
    alert("Failed to load M3U. Please check the URL.");
    console.error(err);
  }
}

function renderCategories() {
  const groups = ["All", "Favorites", ...new Set(channels.map(c => c.group))];
  categoryFilter.innerHTML = groups
    .map(g => `<option value="${g}">${g}</option>`)
    .join("");
}

function renderChannels() {
  const keyword = searchInput.value.toLowerCase();
  const group = categoryFilter.value;
  channelList.innerHTML = "";

  channels.forEach((ch, index) => {
    const isFav = favorites.includes(ch.name);
    if (
      (group === "All" || ch.group === group || (group === "Favorites" && isFav)) &&
      ch.name.toLowerCase().includes(keyword)
    ) {
      const div = document.createElement("div");
      div.className = "channel-card";
      div.innerHTML = `
        <img src="${ch.logo || "https://via.placeholder.com/150"}" alt="${ch.name}">
        <span>${ch.name}</span>
        <div class="fav ${isFav ? "on" : ""}" data-index="${index}">★</div>
      `;
      div.onclick = () => playChannel(index);
      channelList.appendChild(div);
    }
  });
}

function playChannel(index) {
  const ch = channels[index];
  if (!ch) return;

  currentIndex = index;
  nowPlaying.textContent = "Now Playing: " + ch.name;

  if (Hls.isSupported()) {
    const hls = new Hls();
    hls.loadSource(ch.url);
    hls.attachMedia(video);
  } else {
    video.src = ch.url;
  }
}

function toggleFavorite(index) {
  const ch = channels[index];
  if (!ch) return;
  const i = favorites.indexOf(ch.name);
  if (i === -1) {
    favorites.push(ch.name);
  } else {
    favorites.splice(i, 1);
  }
  localStorage.setItem("favorites", JSON.stringify(favorites));
  renderChannels();
}

channelList.addEventListener("click", (e) => {
  if (e.target.classList.contains("fav")) {
    e.stopPropagation();
    toggleFavorite(e.target.dataset.index);
  }
});

searchInput.addEventListener("input", renderChannels);
categoryFilter.addEventListener("change", renderChannels);
importBtn.addEventListener("click", () => {
  const url = customM3U.value.trim();
  if (url) loadM3U(url);
});

document.addEventListener("touchstart", handleSwipeStart, false);
document.addEventListener("touchend", handleSwipeEnd, false);

let xStart = null;

function handleSwipeStart(evt) {
  xStart = evt.touches[0].clientX;
}

function handleSwipeEnd(evt) {
  if (!xStart) return;
  let xEnd = evt.changedTouches[0].clientX;
  if (xEnd - xStart > 50) playChannel((currentIndex - 1 + channels.length) % channels.length);
  else if (xStart - xEnd > 50) playChannel((currentIndex + 1) % channels.length);
  xStart = null;
}

setClock();
loadM3U(m3uUrl);
